### PR TITLE
Expose client details to login view

### PR DESCRIPTION
**Pull request by gavinharriss**
The coded solution to make issue https://github.com/IdentityServer/IdentityServer3/issues/1385 work.

I accidentally made the pull request against master, so this is a re-submit of code against the dev branch instead.

### CODE REVIEW

File: source/Core/Endpoints/AuthenticationController.cs [hunk 1]
```diff
@@ -811,7 +812,10 @@ private void SignOutOfExternalIdP()
                 CurrentUser = context.GetCurrentUserDisplayName(),
                 LogoutUrl = context.GetIdentityServerLogoutUrl(),
                 AntiForgery = antiForgeryToken.GetAntiForgeryToken(),
-                Username = username
+                Username = username,
+                ClientName = client.ClientName,
```
Comment (huysentruitw) [inline source/Core/Endpoints/AuthenticationController.cs]: What if `client == null` ?

Comment (gavinharriss) [inline source/Core/Endpoints/AuthenticationController.cs]: I'd assumed (maybe wrongly) that client would always be present if getting this far. From my limited playing I've never managed to hit the login view without having a valid client present. Without a client I've been redirected to error views.

Comment (huysentruitw) [inline source/Core/Endpoints/AuthenticationController.cs]: I think it doesn't matter if it is always preset, the method is called `FindClientByIdAsync` and not `GetClientByIdAsync`. And the `FindClientByIdAsync` is implemented outside IdSrv3, so you don't have control.

You could formward the `null` value to the string properties or throw a more descriptive exception.

**Comment by gavinharriss** [comment]
I assume it's not worth me changing the code to handle nulls as it is preferred to access these extra properties using an IVIewService instead? Thus the code will not be merged?

**Comment by brockallen** [comment]
We'll take a look this week. Thx.

**Comment by brockallen** [comment]
Ok, looks good. Thanks.

**Pull request merged**